blockstack/gaia
# Feature/cloud api testing

## Goal of PR

Perform integration testing for all drivers https://github.com/blockstack/gaia/issues/197. This PR gives us the following: 
* Assurance that the revocation methodology of depending on cloud API 404s works reliably. 
* Implement more tests for ensuring driver/cloud API compliance and behavioral consistency. 
* Easier to implement/update/test drivers: a pattern for specifying mock driver code and integration config setup code now exists. These are automatically ran against our test suite ensuring driver compliance. 
* A json object for all cloud API credentials are stored in a private CircleCI env var. 

## Implementation

Normalized the method for developers to add to the set of mock driver implementations and the set of driver integration setup methods.

Each driver implementation has an entry for both A) mock instance setup method, and B) an method that instantiates an production driver instance when the credentials/config are specified via env vars.  

Both these sets are iterated over, with each driver being instantiated then ran through the main driver unit test. In additional, several tests dealing with authentication and revocations are now also ran against all configured drivers. 
 
## Manual Testing

1. Git pull
1. Specify cloud API credentials using either the `DRIVER_CONFIG_TEST_DATA` env var pointed to the `Driver Credentials JSON` found in the `Gaia Driver Testing Keys` 1password vault (ask @kantai or myself for access). The circleCI tests are with this. 
   * Alternatively, the previous (driver specific) env vars can still be used. For the the list of these env vars see:  https://github.com/blockstack/gaia/blob/65ba7a5d945db1be458b78465bd9ce1a4f6ce223/hub/test/src/testDrivers/integrationTestDrivers.js#L24
    * If no cloud credentials are specified, the integration tests will still be ran using the self-hosted diskDriver and the InMemoryDriver. 
1. `npm i`
1. `npm test`

## Automated Testing

The disk driver now has a mock and self-hosted integration implemented. This, along with the InMemoryDriver, are always ran (no environment/external configuration needed). Addressing @jcnelson's earlier revocation PR comment about using the diskDriver rather than the InMemoryDriver - we can probably remove the InMemoryDriver now that we have this setup. However, this did create the first inter-project npm dev-dependency where the `hub` project now depends the the `reader` project. We should evaluate and confirm that we are OK with this.
